core/commands/dag/dag.go

```diff
@@ -241,3 +278,421 @@ var DagResolveCmd = &cmds.Command{
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		if silent, _ := req.Options[silentOptionName].(bool); silent {
+			req.Options[progressOptionName] = false
+		}
+
+		return nil
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		// !!!NOTE!!! - We take .GCLock, *not* .PinLock. The way the current stack is
+		// built makes it impossible to take both a GC and a Pinning lock, regardless
+		// of orer. Doing so can potentially deadlock if another thread starts GC
+		// ( this is not theoretical - it's been reliably simulated with the insertion
+		// of strategic sleeps()s and hand-triggering GC )
+		//
+		// See api.Pin().Add() vs node.Pinning.Pin() comment further down
+		unlocker := node.Blockstore.GCLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+		showProgress, _ := req.Options[progressOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		var counts ObjectCounts
+		go importWorker(req, &res, &api, &counts, retCh)
+
+		var progressTicker *time.Ticker
+		if showProgress {
+			progressTicker = time.NewTicker(500 * time.Millisecond)
+			defer progressTicker.Stop()
+		}
+
+		var roots *expectedRootsSeen
+
+	ImportLoop:
+		for {
+			select {
+			case ret := <-retCh:
+				if ret.err != nil {
+					return ret.err
+				}
+				roots = ret.roots
+				if progressTicker != nil {
+					progressTicker.Stop()
+					if err := res.Emit(&CarImportOutput{ObjectCounts: counts}); err != nil {
+						return err
+					}
+					os.Stderr.WriteString("\n")
+				}
+				break ImportLoop
+			case <-progressTicker.C:
+				if err := res.Emit(&CarImportOutput{ObjectCounts: counts}); err != nil {
+					return err
+				}
+			case <-req.Context.Done():
+				return req.Context.Err()
+			}
+		}
+
+		// opportunistic pinning: try whatever sticks
+		if doPinRoots {
+
+			for c, seen := range *roots {
+
+				// We need to re-retrieve a block, convert it to ipld, and feed it
+				// to the Pinning interface, sigh...
```

Comment:
We can change the pinning interfaces in a followup patch.

Reply:
Nod, likely not worth it before more ipld-prime stuff lands anyway. I am just leaving comments for the future self.